File: sort/quick_sort/cpp/quicksort.cpp
#include <iostream>
using namespace std;
int partition(int *unsortedArray, int low, int high)
{
    int left = low;
    int pivot = low;
    int right = high;
    while (left < right)
    {

        if (unsortedArray[left] <= unsortedArray[pivot])
        {

            left++;
        }
        if (unsortedArray[right] >= unsortedArray[pivot])
        {

            right--;
        }
        if (unsortedArray[left] > unsortedArray[right])
        {

            int c = unsortedArray[left] + unsortedArray[right];
            unsortedArray[right] = c - unsortedArray[right];
            unsortedArray[left] = c - unsortedArray[left];
        }
    }
    int c = unsortedArray[right] + unsortedArray[pivot];
    unsortedArray[right] = c - unsortedArray[right];
    unsortedArray[pivot] = c - unsortedArray[pivot];

    return right;
}
int *quickSort(int *unsortedArray, int low, int high)
{

    if (low < high)
    {
        int partitionValue = partition(unsortedArray, low, high);
        quickSort(unsortedArray, low, partitionValue - 1);
        quickSort(unsortedArray, partitionValue + 1, high);
    }
    return unsortedArray;
}
int main()
{
    int n;
    cout << "enter number of elements" << endl;
    cin >> n;
    int unsortedArray[n];
    for (int i = 0; i < n; i++)
    {
        cin >> unsortedArray[i];
    }
    int *sortedArray = quickSort(unsortedArray, 0, n - 1);
    for (int i = 0; i < n; i++)
    {
        cout << sortedArray[i] << endl;
    }
}

File: sort/quick_sort/go/quick_sort.go
// Quick Sort in Golang
package main

import (
	"fmt"
	"math/rand"
)

func main() {
	var n int

	fmt.Printf("Enter number of integers : ")
	fmt.Scan(&n)

	arr := make([]float64, n)

	fmt.Printf("\nEnter array elements : ")

	for i := 0; i < n; i++ {
		fmt.Scan(&arr[i])
	}

	quicksort(arr)
	fmt.Printf("Sorted array is ")

	for i := 0; i < n; i++ {
		fmt.Printf("%v ", arr[i])
	}
}

func quicksort(a []float64) []float64 {
	if len(a) < 2 {
		return a
	}

	left, right := 0, len(a)-1

	pivot := rand.Int() % len(a)

	a[pivot], a[right] = a[right], a[pivot]

	for i, _ := range a {
		if a[i] < a[right] {
			a[left], a[i] = a[i], a[left]
			left++
		}
	}

	a[left], a[right] = a[right], a[left]

	quicksort(a[:left])
	quicksort(a[left+1:])

	return a
}
